internal/remotefile/resolver.go

```diff
@@ -0,0 +1,68 @@
+package remotefile
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/osbuild/osbuild-composer/internal/worker"
+)
+
+type resolveResult struct {
+	url     string
+	content []byte
+	err     error
+}
+
+// TODO: could make this more generic
+// since this is shared with the container
+// resolver
+type Resolver struct {
+	jobs  int
+	queue chan resolveResult
+
+	ctx context.Context
+}
+
+func NewResolver() Resolver {
+	return Resolver{
+		ctx:   context.Background(),
+		queue: make(chan resolveResult, 2),
+	}
+}
+
+func (r *Resolver) Add(url string) {
+	client := NewClient()
+	r.jobs += 1
+
+	go func() {
+		content, err := client.Resolve(url)
+		r.queue <- resolveResult{url: url, content: content, err: err}
+	}()
+}
+
+func (r *Resolver) Finish() ([]worker.FileResolveJobResultItem, error) {
+
+	resultItems := make([]worker.FileResolveJobResultItem, 0, r.jobs)
+	errs := make([]string, 0, r.jobs)
+	for r.jobs > 0 {
+		result := <-r.queue
+		r.jobs -= 1
+
+		if result.err != nil {
+			errs = append(errs, result.err.Error())
+		} else {
+			resultItems = append(resultItems, worker.FileResolveJobResultItem{
+				URL:     result.url,
+				Content: result.content,
+			})
+		}
+	}
+
+	if len(errs) > 0 {
+		detail := strings.Join(errs, "; ")
+		return nil, fmt.Errorf(detail)
+	}
```

Comment:
This block got me thinking if it would make sense to connect the `FileResolveJobResultItem` with the error that could have happened. Especially the URL. My thinking is that the error may not contain the URL that failed to resolve and as a result it will be hard to determine what actually failed.

I was thinking about potentially embedding the error itself in the `FileResolveJobResultItem` structure. You could then add a variant of 
https://github.com/osbuild/osbuild-composer/blob/f8b39cd34efd175fbb59a9c0b08072e4e2a951d9/internal/worker/json.go#L59-L76
to collect errors from the slice of `FileResolveJobResultItem` and use it in the job implementation to decide if the resolution was successful.

Alternatively, you could add a new type implementing the `Error` interface, which would hold the error and then you could just return `[]error`.

Something like:

```go
type ResolutionError struct {
	URL string
	Err error
}

func (e *ResolutionError) Error() string {
	return fmt.Sprintf("%s: %v", e.URL, e.Err)
}
```

I think that we are potentially loosing some information by just concatenating all errors into single one and returning it.

Nevertheless I checked the container resolver and it has the same behavior (which I don't like)...

Reply:
Cool, I've added a variant of this - that is if I understood this comment correctly. It may need to be fleshed out a little bit more (or possibly some additional tests might be needed)